collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java

```diff
@@ -594,46 +597,22 @@ private void loadFromIntent(Intent intent) {
             }
 
             instancePath = formInfo.getInstancePath();
+            List<Form> candidateForms = formsRepository.getAll(formInfo.getFormId(), formInfo.getFormVersion());
 
-            String jrFormId = formInfo.getFormId();
-            String jrVersion = formInfo.getFormVersion();
-
-            String[] selectionArgs;
-            String selection;
-            if (jrVersion == null) {
-                selectionArgs = new String[]{jrFormId};
-                selection = FormsColumns.JR_FORM_ID + "=? AND "
-                        + FormsColumns.JR_VERSION + " IS NULL";
-            } else {
-                selectionArgs = new String[]{jrFormId, jrVersion};
-                selection = FormsColumns.JR_FORM_ID + "=? AND "
-                        + FormsColumns.JR_VERSION + "=?";
-            }
-
-            int formCount = FormsDaoHelper.getFormsCount(selection, selectionArgs);
-            if (formCount < 1) {
+            if (candidateForms.size() < 1) {
                 createErrorDialog(getString(
                         R.string.parent_form_not_present,
-                        jrFormId)
-                                + ((jrVersion == null) ? ""
-                                : "\n"
-                                + getString(R.string.version)
-                                + " "
-                                + jrVersion),
+                        formInfo.getFormId())
+                                + ((formInfo.getFormVersion() == null) ? ""
+                                : "\n" + getString(R.string.version) + " " + formInfo.getFormVersion()),
                         EXIT);
                 return;
-            } else {
-                formPath = FormsDaoHelper.getFormPath(selection, selectionArgs);
-
-                /**
-                 * Still take the first entry, but warn that there are multiple rows. User will
-                 * need to hand-edit the SQLite database to fix it.
-                 */
-                if (formCount > 1) {
-                    createErrorDialog(getString(R.string.survey_multiple_forms_error), EXIT);
-                    return;
-                }
+            } else if (candidateForms.stream().filter(f -> !f.isDeleted()).count() > 1) {
```

Comment:
So we can't just use `FormsRepository#getAllNotDeleted` because we want a different error message here?

Reply:
(@seadowg and I talked about this but answering for posterity and to double check my thinking)

That's right -- the first error case is that there are no forms at all (deleted or not) that match. The second error case is if there are more than one **non-deleted** that match the formid/version. Basically:
- we're happy to use a soft-deleted form if that's the only one that matches
- we're happy to have formid/version matches as long as exactly one is non-deleted

So initially we need to fetch all forms including soft-deleted ones. Then I could have gone back to the db with getAllNotDeleted but filtering the list seemed simpler/more performant.